Reintroduce spotlight / "important traits" feature

Manishearth opened the pull request:
(Reopened version of https://github.com/rust-lang/rust/pull/74111 because Github is broken, see discussion there)


Fixes https://github.com/rust-lang/rust/issues/73785

This PR reintroduces the "spotlight" ("important traits") feature.

A couple changes have been made:

As there were concerns about its visibility, it has been moved to be next to the return type, as opposed to being on the side.


It also no longer produces a modal, it shows the traits on hover, and it can be clicked on to pin the hover bubble.

![image](https://user-images.githubusercontent.com/1617736/86674555-a82d2600-bfad-11ea-9a4a-a1a9ffd66ae5.png)

![image](https://user-images.githubusercontent.com/1617736/86674533-a1061800-bfad-11ea-9e8a-c62ad86ed0d7.png)


It also works fine on mobile:

![image](https://user-images.githubusercontent.com/1617736/86674638-bda25000-bfad-11ea-8d8d-1798b608923e.png)

Manishearth committed:
Revert "Remove "important traits" feature"

This reverts commit 1244ced9580b942926afc06815e0691cf3f4a846.

Manishearth committed:
Revert "Remove spotlight usage"

This reverts commit 13c6d5819aae3c0de6a90e7f17ea967bf4487cbb.

Manishearth committed:
Move spotlight next to the return type

Manishearth committed:
Make spotlight show on hover

This makes the spotlight show on hover instead of click. Clicks can be
used to persist it, which is also what's used on mobile.

Manishearth committed:
Review comments for JS

rust-highfive commented:
Some changes occurred in HTML/CSS/JS.

cc @GuillaumeGomez

Some changes occurred in HTML/CSS themes.

cc @GuillaumeGomez

rust-highfive commented:
r? @ollie27

(rust_highfive has picked a reviewer for you, use r? to override)

Manishearth commented:
r? @GuillaumeGomez

Manishearth committed:
Add Ayu theme for spotlight

GuillaumeGomez commented on src/librustdoc/html/static/rustdoc.css:
```diff
@@ -1097,9 +1096,46 @@ h3 > .collapse-toggle, h4 > .collapse-toggle {
+	/* Themes will override the color here,
+	   but it's necessary to make it opaque */
+	background-color: #eee;
+	border: 1px solid #999;
```
Please do not set the border color here but in the themes.

Manishearth committed:
Remove !important on border-color and background-color

ollie27 commented on src/librustdoc/html/render.rs:
```diff
@@ -3558,6 +3560,62 @@ fn should_render_item(item: &clean::Item, deref_mut_: bool) -> bool {
+
+    if let Some(did) = decl.output.def_id() {
+        let c = cache();
+        if let Some(impls) = c.impls.get(&did) {
```
This isn't a good way to find applicable impls because it doesn't take into account things like generic parameters and mutability. For example:

![image](https://user-images.githubusercontent.com/7189418/87577858-14e9a400-c6c3-11ea-8e5f-14ca6fd8caf1.png)

and

![image](https://user-images.githubusercontent.com/7189418/87577906-2632b080-c6c3-11ea-9b55-bd5014915a01.png)

I'd imagine this will need to use a mechanism similar to how auto trait and blankets impls are found.

Manishearth replied:
Given that this is reintroducing an old feature that was previously accepted in this form, can we handle this as a followup?

Manishearth replied:
Like, I want to ladn this so we can gather more community feedback on this feature and grow it. It's a feature that existed in this form for a while and was recently removed, my priority is getting it back in so we can improve it.

ollie27 reviewed (commented):
I've noticed a bit of a graphic glitch for free functions:

![image](https://user-images.githubusercontent.com/7189418/87578615-313a1080-c6c4-11ea-98a2-e8d93553beb2.png)

GuillaumeGomez commented on src/librustdoc/html/render.rs:
```diff
@@ -3558,6 +3560,62 @@ fn should_render_item(item: &clean::Item, deref_mut_: bool) -> bool {
+    if !out.is_empty() {
+        out.insert_str(
+            0,
+            "<span class=\"important-traits\"><span class=\"important-traits-tooltip\">ⓘ<div class='important-traits-tooltiptext'><span class=\"docblock\">"
```
This line is too long, it won't pass tidy check. ;)

Manishearth replied:
Seems to pass tidy

GuillaumeGomez replied:
Oh? I'm surprised but then it's fine. :)

Manishearth commented:
@ollie27 Guillaume has said he's good with this once we have your approval. Would you be okay with landing this with followups filed? I would rather keep this PR scoped to mostly bringing the feature back as is, with some basic UI changes. I want to do improvements differently, with a tracking issue.

ollie27 commented:
I'm really not a fan of reintroducing bugs but I guess it's not the end of the world. There's even an old issue still open (#55082) for the impls that aren't applicable.

Could the issue with free functions I pointed out in https://github.com/rust-lang/rust/pull/74370#pullrequestreview-449195537 be fixed though? That wasn't an issue with the old implementation. If it's too difficult to fix then the important traits feature could be disabled for those functions for now.

Manishearth commented:
Yeah I can fix that, i didn't immediately understand what the bug was but I do now

Manishearth committed:
Don't position:relative on all pres

We need it for run buttons (https://github.com/rust-lang/rust/pull/44671), but not function defs

Manishearth commented:
@ollie27 It was the selector in https://github.com/rust-lang/rust/pull/44671 not being specific enough, Fixed

Manishearth commented:
![image](https://user-images.githubusercontent.com/1617736/87695680-a51bfd80-c744-11ea-98b2-b122c7b3d88c.png)

Manishearth commented:
@ollie27 Guillaume is holding off on r+ing until you confirm the fix above, can you do so?

bors commented:
:umbrella: The latest upstream changes (presumably #72481) made this pull request unmergeable. Please resolve the merge conflicts.

ollie27 commented:
Looks good now, glad it was an easy fix.

Manishearth commented:
@bors r=GuillaumeGomez

bors commented:
:pushpin: Commit c621a54eeb71248927dff8ac0c76f279cf756f28 has been approved by `GuillaumeGomez`

<!-- @bors r=GuillaumeGomez c621a54eeb71248927dff8ac0c76f279cf756f28 -->
<!-- homu: {"type":"Approved","sha":"c621a54eeb71248927dff8ac0c76f279cf756f28","approver":"GuillaumeGomez"} -->

bors commented:
:evergreen_tree: The tree is currently closed for pull requests below priority 5, this pull request will be tested once the tree is reopened

Manishearth commented:
I'll do the follow-ups soonish

Manishearth commented:
Followup: https://github.com/rust-lang/rust/issues/74417

I'll open an internals thread to workshop the UI and name once it's on nightly

Manishearth commented:
Internals thread: https://internals.rust-lang.org/t/feedback-on-important-traits-rustdoc-feature/12752
